File: app-android-biometric-passport-zk/app/src/main/java/org/iranUnchained/feature/voting/logic/VoteSubmissionService.kt
package org.iranUnchained.feature.voting.logic

import android.content.Context
import android.util.Log
import io.reactivex.Observable
import io.reactivex.ObservableEmitter
import org.iranUnchained.R
import org.iranUnchained.base.ActiveConfig
import org.iranUnchained.contracts.ProposalsState
import org.iranUnchained.data.models.IdentityData
import org.iranUnchained.data.models.ProposalData
import org.iranUnchained.data.models.VoteSubmissionAttributes
import org.iranUnchained.data.models.VoteSubmissionRequest
import org.iranUnchained.data.models.VoteSubmissionRequestData
import org.iranUnchained.di.providers.ApiProvider
import org.iranUnchained.logic.persistance.SecureSharedPrefs
import org.iranUnchained.BuildConfig
import org.iranUnchained.utils.CalldataEncoder
import org.iranUnchained.utils.PassportDataExporter
import org.iranUnchained.utils.VoteSMTInputsBuilder
import org.iranUnchained.utils.ZKPTools
import org.iranUnchained.utils.ZKPUseCase
import org.web3j.crypto.Credentials
import org.web3j.crypto.Keys
import org.web3j.tx.RawTransactionManager
import org.web3j.tx.gas.DefaultGasProvider
import java.math.BigInteger
import java.util.Calendar

data class VoteProgress(val step: Int)

class VoteSubmissionService(
    private val context: Context,
    private val apiProvider: ApiProvider
) {
    companion object {
        private const val TAG = "VoteSubmissionService"
    }

    fun submitVote(
        proposalData: ProposalData,
        selectedOptions: List<Int>
    ): Observable<VoteProgress> {
        return Observable.create { emitter ->
            try {
                // Step 0: Building proof inputs
                emitter.onNext(VoteProgress(0))
                val proofInputs = buildProofInputs(proposalData, selectedOptions)

                // Step 1: Generating ZK proof
                emitter.onNext(VoteProgress(1))
                val zkProof = generateGroth16Proof(proofInputs)

                if (BuildConfig.DEBUG) {
                    PassportDataExporter.exportProofData(
                        zkProof,
                        proofInputs.registrationRoot,
                        proofInputs.currentDate,
                        proofInputs.proposalEventId,
                        proofInputs.nullifier,
                        proofInputs.citizenship,
                        proofInputs.identityCreationTimestamp,
                        proofInputs.votes
                    )
                    Log.i(TAG, "Proof data exported to logcat for test extraction")
                }

                // Step 2: Submitting vote
                emitter.onNext(VoteProgress(2))
                val txHash = submitToRelayer(
                    proposalData,
                    proofInputs,
                    zkProof,
                    selectedOptions
                )
                Log.i(TAG, "Vote submitted, tx: $txHash")

                // Save voted state
                val identityData = getIdentityData()
                if (identityData != null) {
                    SecureSharedPrefs.saveVotedAddress(
                        context,
                        identityData.nullifierHex,
                        proposalData.votingContractAddress
                    )
                }

                // Step 3: Vote confirmed
                emitter.onNext(VoteProgress(3))
                emitter.onComplete()
            } catch (e: Exception) {
                Log.e(TAG, "Vote submission failed", e)
                emitter.onError(e)
            }
        }
    }

    private data class ProofInputs(
        val registrationRoot: ByteArray,
        val currentDate: BigInteger,
        val proposalEventId: BigInteger,
        val nullifier: BigInteger,
        val citizenship: BigInteger,
        val identityCreationTimestamp: BigInteger,
        val votes: List<BigInteger>,
        val inputsJson: String
    )

    private fun buildProofInputs(
        proposalData: ProposalData,
        selectedOptions: List<Int>
    ): ProofInputs {
        val identityData = getIdentityData()
            ?: throw IllegalStateException("Identity not found. Please scan your passport first.")

        val web3j = apiProvider.web3
        val ecKeyPair = Keys.createEcKeyPair()
        val credentials = Credentials.create(ecKeyPair)
        val gasProvider = DefaultGasProvider()

        // Get proposal event ID for nullifier computation
        val proposalsState = ProposalsState.load(
            ActiveConfig.PROPOSAL_ADDRESS, web3j, credentials, gasProvider
        )
        val proposalEventId = proposalsState.getProposalEventId(
            BigInteger.valueOf(proposalData.proposalId)
        ).send()

        // Get registration SMT root
        val registrationRoot = getRegistrationRoot(credentials)

        // Encode current date as 6 ASCII bytes (YYMMDD)
        val cal = Calendar.getInstance()
        val currentDate = CalldataEncoder.encodeDateAsAsciiBytes(
            cal.get(Calendar.YEAR),
            cal.get(Calendar.MONTH) + 1,
            cal.get(Calendar.DAY_OF_MONTH)
        )

        // Encode votes: single-element bitmask array [1 << selectedOptionIndex]
        val votes = CalldataEncoder.encodeVoteBitmasks(selectedOptions, proposalData.options.size)

        // Parse identity fields
        val nullifier = BigInteger(identityData.nullifierHex, 16)
        val citizenship = getCitizenshipCode()
        val identityCreationTimestamp = BigInteger(identityData.timeStamp)

        // Build circuit inputs JSON
        val rootHex = org.web3j.utils.Numeric.toHexStringNoPrefix(registrationRoot)
        val inputsJson = VoteSMTInputsBuilder.buildJson(
            registrationRootHex = rootHex,
            currentDate = currentDate,
            proposalEventId = proposalEventId,
            nullifier = nullifier,
            secretKey = BigInteger(identityData.secretKeyHex, 16),
            citizenship = citizenship,
            identityCreationTimestamp = identityCreationTimestamp,
            votes = votes,
            proposalId = BigInteger.valueOf(proposalData.proposalId)
        )

        return ProofInputs(
            registrationRoot = registrationRoot,
            currentDate = currentDate,
            proposalEventId = proposalEventId,
            nullifier = nullifier,
            citizenship = citizenship,
            identityCreationTimestamp = identityCreationTimestamp,
            votes = votes,
            inputsJson = inputsJson
        )
    }

    private fun generateGroth16Proof(inputs: ProofInputs): org.iranUnchained.data.models.ZkProof {
        if (!BuildConfig.USE_REAL_PROOFS) {
            Log.i(TAG, "Mock proofs enabled: generating random proof (VerifierMock accepts any proof)")
            return generateMockProof()
        }

        Log.i(TAG, "Generating real Groth16 proof via vote_smt circuit...")
        val zkpUseCase = ZKPUseCase(context)
        val zkpTools = ZKPTools(context)

        return zkpUseCase.generateZKP(
            R.raw.vote_smt_zkey,
            R.raw.vote_smt,
            inputs.inputsJson.toByteArray(),
            zkpTools::voteSMT
        )
    }

    private fun generateMockProof(): org.iranUnchained.data.models.ZkProof {
        // Generate random proof points as decimal strings — VerifierMock accepts any proof
        val rand = {
            val r = java.security.SecureRandom()
            val b = ByteArray(32); r.nextBytes(b)
            BigInteger(1, b).toString() // decimal string (CalldataEncoder uses BigInteger(str))
        }
        val proof = org.iranUnchained.data.models.Proof(
            pi_a = listOf(rand(), rand()),
            pi_b = listOf(listOf(rand(), rand()), listOf(rand(), rand())),
            pi_c = listOf(rand(), rand()),
            protocol = "groth16"
        )
        return org.iranUnchained.data.models.ZkProof(
            proof = proof,
            pub_signals = emptyList()
        )
    }

    private fun submitToRelayer(
        proposalData: ProposalData,
        proofInputs: ProofInputs,
        zkProof: org.iranUnchained.data.models.ZkProof,
        selectedOptions: List<Int>
    ): String {
        // Encode userPayload
        val userPayload = CalldataEncoder.encodeUserPayload(
            proposalId = BigInteger.valueOf(proposalData.proposalId),
            votes = proofInputs.votes,
            nullifier = proofInputs.nullifier,
            citizenship = proofInputs.citizenship,
            identityCreationTimestamp = proofInputs.identityCreationTimestamp
        )

        // Encode full execute() calldata
        val calldata = CalldataEncoder.encodeExecuteCalldata(
            registrationRoot = proofInputs.registrationRoot,
            currentDate = proofInputs.currentDate,
            userPayload = userPayload,
            proof = zkProof.proof
        )

        if (BuildConfig.IS_LOCAL_DEV) {
            val payloadHex = org.web3j.utils.Numeric.toHexStringNoPrefix(userPayload)
            Log.i(TAG, "Local dev: userPayload=$payloadHex")
            Log.i(TAG, "Local dev: calldata=$calldata")
            return submitDirectToChain(proposalData.votingContractAddress, calldata)
        }

        // Submit to relayer
        val request = VoteSubmissionRequest(
            data = VoteSubmissionRequestData(
                attributes = VoteSubmissionAttributes(
                    tx_data = calldata,
                    destination = proposalData.votingContractAddress
                )
            )
        )

        val response = apiProvider.circuitBackend
            .submitVote(ActiveConfig.VOTE_LINK, request)
            .blockingGet()

        return response.data.id
    }

    /**
     * Submit vote transaction directly to the local Hardhat chain,
     * bypassing the relayer gateway (which requires Docker services).
     * Uses Hardhat's well-known account #0 to send the transaction.
     */
    private fun submitDirectToChain(votingContractAddress: String, calldata: String): String {
        Log.i(TAG, "Local dev: submitting tx directly to chain (bypassing relayer)")
        Log.i(TAG, "Local dev: target=$votingContractAddress selector=${calldata.take(10)} calldataLen=${calldata.length}")
        val web3j = apiProvider.web3

        // Hardhat account #0 — well-known test private key (no real funds)
        val credentials = Credentials.create(
            "ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"
        )

        val txManager = RawTransactionManager(web3j, credentials, 31337L)
        val response = txManager.sendTransaction(
            DefaultGasProvider.GAS_PRICE,
            DefaultGasProvider.GAS_LIMIT,
            votingContractAddress,
            calldata,
            BigInteger.ZERO
        )

        if (response.hasError()) {
            throw RuntimeException("Transaction failed: ${response.error.message}")
        }

        Log.i(TAG, "Local dev: tx hash = ${response.transactionHash}")
        return response.transactionHash
    }

    private fun getIdentityData(): IdentityData? {
        val json = SecureSharedPrefs.getIdentityData(context) ?: return null
        return IdentityData.fromJson(json)
    }

    private fun getRegistrationRoot(credentials: Credentials): ByteArray {
        val web3j = apiProvider.web3
        val gasProvider = DefaultGasProvider()

        val registration = org.iranUnchained.contracts.SRegistration.load(
            ActiveConfig.REGISTRATION_ADDRESS, web3j, credentials, gasProvider
        )
        return registration.getRoot().send()
    }

    private fun getCitizenshipCode(): BigInteger {
        val issuerAuthority = SecureSharedPrefs.getIssuerAuthority(context)
        return if (issuerAuthority != null) {
            // Convert 3-letter country code to numeric representation
            val bytes = issuerAuthority.toByteArray(Charsets.US_ASCII)
            BigInteger(1, bytes)
        } else {
            BigInteger.ZERO
        }
    }

}

File: app-ios-biometric-passport-zk/IranUnchained/Code/Features/Voting/Services/VoteSubmissionService.swift
//
//  VoteSubmissionService.swift
//  IranUnchained
//
//  Orchestrates the full vote submission flow — direct port of Android's VoteSubmissionService.kt.
//
//  Steps:
//    0: Build proof inputs (fetch SMT root, proposalEventId, encode date/votes)
//    1: Generate ZK proof (or mock proof for local dev)
//    2: Encode calldata + submit (direct-to-chain for local dev, relayer for production)
//    3: Confirmed
//

import Foundation
import OSLog

class VoteSubmissionService {
    private let config: Config
    private let user: User?
    private let logger = Logger(subsystem: "org.IranUnchained", category: "VoteSubmission")

    init(config: Config, user: User?) {
        self.config = config
        self.user = user
    }

    func submitVote(
        proposal: ProposalData,
        selectedOptions: [Int],
        onProgress: @escaping (Int) -> Void
    ) async throws {
        // Step 0: Build proof inputs
        onProgress(0)
        logger.info("Vote submission: selectedOptions=\(selectedOptions) (0-indexed)")
        let proofInputs = try await buildProofInputs(proposal: proposal, selectedOptions: selectedOptions)

        // Step 1: Generate ZK proof
        onProgress(1)
        let proof: ZKProofPoints
        if config.isLocalDev {
            logger.info("Mock proofs enabled: generating random proof (VerifierMock accepts any proof)")
            proof = ZKProofPoints.mock()
        } else {
            // TODO: Integrate real Groth16 proof generation via ZKUtils
            throw "Real ZK proof generation not yet implemented on iOS. Use local dev build."
        }

        // Step 2: Submit vote
        onProgress(2)
        let txHash = try await submitToChain(
            proposal: proposal,
            proofInputs: proofInputs,
            proof: proof
        )
        logger.info("Vote submitted, tx: \(txHash)")

        // Step 3: Confirmed
        onProgress(3)
    }

    // MARK: - Build Proof Inputs

    private struct ProofInputs {
        let registrationRoot: String      // 64-char hex (no 0x)
        let currentDate: String           // hex-encoded ASCII date
        let proposalEventId: String       // 64-char hex
        let nullifier: String             // hex
        let citizenship: String           // hex
        let identityCreationTimestamp: String // hex
        let votes: [UInt64]
        let inputsJSON: String            // circuit inputs
    }

    private func buildProofInputs(
        proposal: ProposalData,
        selectedOptions: [Int]
    ) async throws -> ProofInputs {
        guard let user = user else {
            throw "Identity not found. Please scan your passport first."
        }

        let rpcURL = config.rarimo.targetChainRPCURL

        // Get proposal event ID
        let proposalEventIdHex = try await ProposalProvider.getProposalEventId(
            rpcURL: rpcURL,
            contractAddress: config.freedom.proposalsStateAddress,
            proposalId: proposal.proposalId
        )

        // Get registration SMT root
        let registrationRoot = try await ProposalProvider.getRegistrationRoot(
            rpcURL: rpcURL,
            contractAddress: config.freedom.registrationContractAddress
        )

        // Encode current date as ASCII bytes
        let cal = Calendar.current
        let now = Date()
        let currentDate = CalldataEncoder.encodeDateAsAsciiBytes(
            year: cal.component(.year, from: now),
            month: cal.component(.month, from: now),
            day: cal.component(.day, from: now)
        )

        // Encode votes
        let votes = CalldataEncoder.encodeVoteBitmasks(selectedOptions: selectedOptions, totalOptions: proposal.options.count)
        logger.info("Vote bitmask: \(votes[0]) (binary: \(String(votes[0], radix: 2)))")

        // Identity fields
        let nullifierHex = hexFromSecretKey(user.secretKey)
        let citizenshipHex = citizenshipToHex(user.issuingAuthority)
        let timestampHex = String(user.creationTimestamp, radix: 16)

        // Circuit inputs (Semaphore-style)
        let voteValue = votes.first ?? 0
        let pathElements = [String](repeating: "0", count: 20)
        let pathIndices = [String](repeating: "0", count: 20)

        let inputsJSON = VoteSMTInputsBuilder.buildJSON(
            root: registrationRoot,
            nullifierHash: nullifierHex, // Phase 1: nullifierHash = nullifier
            nullifier: nullifierHex,
            vote: String(voteValue),
            secret: nullifierHex,
            pathElements: pathElements,
            pathIndices: pathIndices
        )

        return ProofInputs(
            registrationRoot: registrationRoot,
            currentDate: currentDate,
            proposalEventId: proposalEventIdHex,
            nullifier: nullifierHex,
            citizenship: citizenshipHex,
            identityCreationTimestamp: timestampHex,
            votes: votes,
            inputsJSON: inputsJSON
        )
    }

    // MARK: - Submit

    private func submitToChain(
        proposal: ProposalData,
        proofInputs: ProofInputs,
        proof: ZKProofPoints
    ) async throws -> String {
        // Encode userPayload
        let userPayload = CalldataEncoder.encodeUserPayload(
            proposalId: UInt64(proposal.proposalId),
            votes: proofInputs.votes,
            nullifier: proofInputs.nullifier,
            citizenship: proofInputs.citizenship,
            identityCreationTimestamp: proofInputs.identityCreationTimestamp
        )

        // Encode full calldata
        let calldata = CalldataEncoder.encodeExecuteCalldata(
            registrationRoot: proofInputs.registrationRoot,
            currentDate: proofInputs.currentDate,
            userPayload: userPayload,
            proof: proof
        )

        if config.isLocalDev {
            return try await submitDirectToChain(
                votingContractAddress: proposal.votingContractAddress,
                calldata: calldata
            )
        } else {
            return try await submitToRelayer(
                votingContractAddress: proposal.votingContractAddress,
                calldata: calldata
            )
        }
    }

    /// Submit directly to local Hardhat chain using account #0.
    private func submitDirectToChain(
        votingContractAddress: String,
        calldata: String
    ) async throws -> String {
        logger.info("Local dev: submitting tx directly to chain")
        let rpc = RawRPCClient(rpcURL: config.rarimo.targetChainRPCURL)

        // Use Hardhat account #0 via eth_sendTransaction (Hardhat auto-signs for known accounts)
        let hardhatAccount0 = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"

        let txParams: [String: Any] = [
            "from": hardhatAccount0,
            "to": votingContractAddress,
            "data": calldata,
            "gas": "0x1e8480" // 2,000,000
        ]

        let txHash = try await rpc.sendTransaction(params: txParams)
        logger.info("Local dev: tx hash = \(txHash)")
        return txHash
    }

    /// Submit to the proof-verification-relayer for production.
    private func submitToRelayer(
        votingContractAddress: String,
        calldata: String
    ) async throws -> String {
        let relayerURL = config.freedom.proofVerificationRelayerURL
        var requestURL = relayerURL
        requestURL.append(path: "/integrations/proof-verification-relayer/v1/vote")

        let body: [String: Any] = [
            "data": [
                "tx_data": calldata,
                "destination": votingContractAddress
            ]
        ]

        var request = URLRequest(url: requestURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)

        if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode != 200 {
            let responseBody = String(data: data, encoding: .utf8) ?? ""
            throw "Relayer error (HTTP \(httpResponse.statusCode)): \(responseBody)"
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let dataObj = json["data"] as? [String: Any],
              let txHash = dataObj["id"] as? String
        else {
            throw "Invalid relayer response"
        }

        return txHash
    }

    // MARK: - Helpers

    /// Derive a hex nullifier from the user's secret key.
    /// For local dev, just use the first 32 bytes of the key as hex.
    private func hexFromSecretKey(_ secretKey: String) -> String {
        // secretKey is the BJJ secret key string from Go Identity lib
        // Convert to hex representation for use in circuit inputs
        let bytes = Array(secretKey.utf8)
        if bytes.count >= 32 {
            return bytes.prefix(32).map { String(format: "%02x", $0) }.joined()
        }
        // If shorter, pad with zeros
        return bytes.map { String(format: "%02x", $0) }.joined()
            + String(repeating: "00", count: max(0, 32 - bytes.count))
    }

    /// Convert 3-letter country code to hex (same as Android's getCitizenshipCode).
    private func citizenshipToHex(_ issuingAuthority: String) -> String {
        let bytes = Array(issuingAuthority.utf8)
        return bytes.map { String(format: "%02x", $0) }.joined()
    }
}
